cpp/tests/rolling/rolling_test.cpp

```diff
@@ -0,0 +1,723 @@
+    std::vector<cudf::size_type> ref_data(num_rows);
+    std::vector<bool> ref_valid(num_rows);
+
+    // input data and mask
+  
+    std::vector<bitmask_type> in_valid; 
+    //std::tie(in_col, in_valid) = cudf::test::to_host<T>(input); 
+    in_valid = cudf::test::bitmask_to_host(input);
+    bitmask_type* valid_mask = in_valid.data();
+    
+    for(size_type i = 0; i < num_rows; i++) {
+      size_type count = 0;
+      // load sizes
+      min_periods = std::max(min_periods, 1); // at least one observation is required
+
+      // compute bounds
+      auto window = window_col[i%window_col.size()];
+      auto forward_window = forward_window_col[i%forward_window_col.size()];
+      size_type start_index = std::max((size_type)0, i - window + 1);
+      size_type end_index = std::min(num_rows, i + forward_window + 1); // exclusive
+      
+      // aggregate
+      for (size_type j = start_index; j < end_index; j++) {
+        if (!input.nullable() || cudf::bit_is_set(valid_mask, j))
+          count++;
+      }
+
+      ref_valid[i] = (count >= min_periods);
+      if (ref_valid[i])
+        ref_data[i] = count;
+    }
+
+    fixed_width_column_wrapper<cudf::size_type> col(ref_data.begin(), ref_data.end(),
+                                                    ref_valid.begin());
+    return col.release();
+  }
+
+  template<typename agg_op, bool is_mean,
+           std::enable_if_t<cudf::detail::is_supported<T, agg_op, is_mean>()>* = nullptr>
+  std::unique_ptr<cudf::column> create_reference_output(cudf::column_view const& input,
+                                                        std::vector<size_type> const& window_col,
+                                                        std::vector<size_type> const& forward_window_col,
+                                                        size_type min_periods)
+  {
+    size_type num_rows = input.size();
+    std::vector<T> ref_data(num_rows);
+    std::vector<bool> ref_valid(num_rows);
+
+    // input data and mask
+    std::vector<T> in_col;
+    std::vector<bitmask_type> in_valid; 
+    std::tie(in_col, in_valid) = cudf::test::to_host<T>(input); 
+    bitmask_type* valid_mask = in_valid.data();
+    
+    agg_op op;
+    for(size_type i = 0; i < num_rows; i++) {
+      T val = agg_op::template identity<T>();
+      size_type count = 0;
+      // load sizes
+      min_periods = std::max(min_periods, 1); // at least one observation is required
+
+      // compute bounds
+      auto window = window_col[i%window_col.size()];
+      auto forward_window = forward_window_col[i%forward_window_col.size()];
+      size_type start_index = std::max((size_type)0, i - window + 1);
+      size_type end_index = std::min(num_rows, i + forward_window + 1); // exclusive
+      
+      // aggregate
+      for (size_type j = start_index; j < end_index; j++) {
+        if (!input.nullable() || cudf::bit_is_set(valid_mask, j)) {
+          val = op(in_col[j], val);
+          count++;
+        }
+      }
```

Comment:
Same comment as above.

Reply:
In this case there are two things in the loop -- the op() and the count. So moving the input.nullable outside the loop would cause me to have to have two loops.